README.md

```diff
@@ -1,7 +1,214 @@
+
+// responsible for decoding a user
+// similar to a "toMap" method.
+class UserEncoder implements Deserializable<Map, User> {
+  @override
+  Map to(User input) => <String, dynamic>{
+        'email': input.email,
+        'name': input.name,
+      };
+}
+
+// responsible for encoding a user
+// similar to a "fromMap" method.
+class UserDecoder implements Serializable<Map, User> {
+  @override
+  User from(Map input) => User(
+    email: input['email'] as String,
+    name: input['name'] as String,
+  );
+}
+```
+
+
+
+### Interceptors
+
+Interceptors give you the ability to transform, change or listen to request or response calls.
+
+This could be something simply like adding an Authorization Header or something more complex like catching and handling a specific error from your api.
+
+1. #### Start by implementing `Interceptor`. 
+
+```dart
+class MyInterceptor implements Interceptor {
+  @override
+  RequestHandler<T> intercept<T>(RequestHandler<T> handler) => (request) async {
+        // adjust the request here using request.copyWith
+        // Maybe add auth headers?
+        // add additional content to the body?
+    
+        // do the call that returns the response
+        final response = await handler(request);
+      
+        // do additional response handling
+        // maybe resend if the response failed because of an expired token?
+        // maybe log out the user if he has no permission to this endpoint?
+        return response;
+      };
+}
+```
+
+An example could look like this:
+
+```dart
+class AuthInterceptor implements Interceptor {
+  @override
+  RequestHandler<T> intercept<T>(RequestHandler<T> handler) => (request) async {
+        final updatedRequest = request.copyWith(
+          headers: {"Authorization": await getToken()},
+        );
+
+	return handler(updatedRequest);
+  };
+}
+```
+
+or
+
+```dart
+class RefreshTokenInterceptor implements Interceptor {
+  @override
+  RequestHandler<T> intercept<T>(RequestHandler<T> handler) => (request) async {
+        final response = await handler(updatedRequest);
+
+        if (!response.isSuccessful && response.errorBody == "Shit! Token expired!") {
+          // refresh the token asynchronously
+          final refreshedRequest = request.copyWith(
+            headers: {"Authorization": await refreshToken()},
+          );
+          
+          // resend the request with the refreshed token!
+          return handler(refreshedRequest);
+        }
+
+        return response;
+  };
+}
+```
+
+
+
+2. #### Chain your interceptors
+
+   You can chain multiple `Interceptor`'s using `Interceptor.fromList`. 
+
+   ```dart
+   final apiProvider = HttpClientApiProvider(
+       httpClient: yourClient,
+       container: yourContainer,
+       interceptor: Interceptor.fromList([
+         TokenInterceptor(),
+         MyInterceptor(),
+         LoggingInterceptor(),
+       ]),
+   )
+   ```
+
+   The given `Interceptors`'s wrap around the source call like an onion.  🤤
```

Comment:
MIssing onion smiley here. Unfortunately it does not exist 😒

Reply:
I knoooooow! There is a smiley for fucking everything but not for an onion 😭